flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala

```diff
@@ -1463,21 +1465,23 @@ class CodeGenerator(
     */
   def addReusableFunction(function: UserDefinedFunction): String = {
     val classQualifier = function.getClass.getCanonicalName
-    val fieldTerm = s"function_${classQualifier.replace('.', '$')}"
+    val functionSerializedData = serialize(function)
+    val fieldTerm =
+      s"""
+         |function_${classQualifier.replace('.', '$')}_${DigestUtils.md5Hex(functionSerializedData)}
```

Comment:
It seems that if we register two identical function (not the same reference), the field term will the same, which will compile error here. Such as:

```
val func1 = new MyFunction
val func2 = new MyFunction
env.register("func1", func1)
env.register("func2", func2)
```

The func1 and func2 is two different object but the serialized bytes are equals.  

BTW, I will suggest to include this case into your new IT cases.